### ISSUE TITLE
Allow client-provided task identifiers

### ISSUE THREAD

**Issue by stefanobaghino** [eremetic-framework/eremetic#184]
I'm opening this ticket to discuss the feasibility of introducing a feature that could be helpful to me, namely, as the title goes, to allow clients to specify the task identifier.

## Context

I'm currently using Eremetic to run tasks scheduled by Airflow (a workflow management tool) on top of Mesos, while having the possibility to specify resource usage at a task-level granularity (something that the native Mesos integration in Airflow does not allow).

Airflow tasks can be defined in Python and arranged together to form a DAG to be executed. A task is considered successful if an operator terminates without throwing, meaning that I needed something that allowed me to run containerized code through Eremetic in a blocking fashion. To achieve this goal and validate the overall feasibility I wrote a toy client (available [here](https://github.com/stefanobaghino/eremetic-synchronous-client)) that just uses the HTTP API of Eremetic to ask for the creation of a task and then polls the returned task identifier until an end condition is met.

## Why this need

Since this first PoC went good, we're trying to productize it. The biggest problem right now is that we've seen things go wrong in the face of network partitions, which is of course completely reasonable due to the hacky nature of my blocking client.

By introducing the possibility for clients to specify a desired task identifier, the client could implement measures to understand if the task being requested is already running or has already run, with the burden of tracking the identifiers falling on the end user. In my use case, we can leverage Airflow to produce unique identifiers.

## Possible implementation decisions

By adding an optional key to the manifest (e.g. `eremetic_desired_task_id`) that when absent makes the identifier assignment default to the current behavior, this feature would also be backward-compatible. If the user provides an identifier that is already in the system, I'd probably fail early and have the HTTP API return a `400 Bad Request`.
I don't see particular problems in allowing both client-provided identifiers and scheduled-provided ones live in a single systems (i.e. having some requests specifying an identifier while others don't).

---

As a further note, I'm not a professional Go developer but perhaps with some review from the authors I could implement the feature myself (if you agree with this request and would not like to personally work on this).

I'd like to have your feedback on this. Thanks!

**Comment by alde** [comment]
I had an idea once of making a pluggable task-id provider, instead of the current implementation, but never got around to it

**Comment by stefanobaghino** [comment]
Maybe I can start working on this and implement the current behavior and the client-provided one as two possible pluggable strategies. Would this be within the project scope?

**Comment by keis** [comment]
I think it would be a good idea to implement this as simple as possible to start with to verify it works as expected, making it pluggable can wait imho.

Do we want this to be the same taskid as we use towards mesos? Otherwise this could perhaps be some sort of "task name" that we internally could let you look up the task by. Less leaky abstraction 😼 

We should look over exactly what we promise about taskids in eremetic specifically around collisions and when recovering through reconcile.

**Comment by stefanobaghino** [comment]
> I think it would be a good idea to implement this as simple as possible to start with to verify it works as expected, making it pluggable can wait imho.

Makes a lot of sense.

> Do we want this to be the same taskid as we use towards mesos? Otherwise this could perhaps be some sort of "task name" that we internally could let you look up the task by. Less leaky abstraction :smirk_cat:

Indeed less leaky. It would also decouple how unique identifiers are handled within Eremetic and how in my use case I need a unique identifier to track a created task in the face of failures, which are instances of the same problem at different layers of abstraction (are they?). This would put the entire burden of defining uniqueness on the client (_if_ they need this property), which still makes sense, I believe.

Perhaps we can calls this `client_ref` or `client_ref_id` (as it serves as reference for the client). The client should assume that, if a task is created with a given `client_ref_id`, the HTTP API can be queried to get the status of all tasks that have been created with that `client_ref_id`. In my use case, having more than one task returned would be an exceptional condition.

> We should look over exactly what we promise about taskids in eremetic specifically around collisions and when recovering through reconcile.

Can you expand on this? Do you refer to existing mechanisms?

---

Thanks for the feedback, @keis. :slightly_smiling_face: I'll have a look at the code and see how this can be done, let me know if you think you can work on this yourself (to prevent overlaps).

**Comment by keis** [comment]
Yeah, I think it makes sense calling it the same problem but different layer and in particular we have promise as part of the contract between eremetic <-> mesos to keep the taskid unique. But the question if we can use it directly would be if we can extend this contract to your usecase. E.g can we assume you would want to do retries using the same ID if things goes wrong.

About the naming I completely forgot there already is a task.Name we could use for this. Not in the api though. Do you think this would make sense? I also like the idea of opening this up to return multiple results as we could then add this to ListTasks with a query parameter to select by name/ref.

>>We should look over exactly what we promise about taskids in eremetic specifically around collisions and when recovering through reconcile.

>Can you expand on this? Do you refer to existing mechanisms?
Yes, so currently we pretty much rely on uuid being globally unique enough for everything. But if we let the id be specified from the outside where it could be reused things could get weird.

Consider a case like this
1, you generate an id and start a new eremetic task
2, bad things happen, eremetic server dies, and a new one starts
3, you re-submit with the same id and eremetic queues it
4, we restore running task from mesos and the same id shows up.
5, ...
6, wut?

Another things with server restarts to consider is that we might tell you nothing of that name exists at first and later change our mind when have reconciled with the running tasks of mesos. Using a external database for the tasks should solve this though. Only zookeeper right now but the db backend is pluggable.

**Comment by zmalik** [comment]
We are also looking forward to being able to know the state of a **job** that can have one or more execution using its task.Name

I think this a perfect example of leveraging task.Name. And if we use user-provided task IDs could be a problem. As it could lead people to reuse taskIDs to identify different executions of same job. That could limit in future flexibility of eremetic to develop some basic jobs related features. Just to say something, retries done by eremetic, execute a job and if it fails retry it **n** times. And each time the task ID would be different as they are different executions. 

Long story short, I agree 100% that task Ids should be unique.

Moreover, it is strongly discouraged by Mesos to reuse a task ID.
[MESOS-2198](https://issues.apache.org/jira/browse/MESOS-2198)

**Comment by stefanobaghino** [comment]
@keis
> About the naming I completely forgot there already is a task.Name we could use for this. Not in the api though. Do you think this would make sense? I also like the idea of opening this up to return multiple results as we could then add this to ListTasks with a query parameter to select by name/ref.

So, you what you would suggest is:
- allowing non-unique `task.Name`s to be provided by the user
- adding an HTTP end point to query status by `task.Name`
- if this is 

If I got it correctly, it doesn't sound bad to me, although it's important to document how `task.Name` could serve as a non-unique reference that can point at different tasks (just to ensure that users do not mistakenly assume that `task.Name` bears some significance of identity). 

@zmalik
> Long story short, I agree 100% that task Ids should be unique.

I do agree with the point, my initial idea of allowing client-provided identifiers was to immediately refuse to create a task if the provided identifier was non-unique. So, yes, I absolutely support this. :slightly_smiling_face:

**Comment by keis** [comment]
Yes, and it sounds like this would solve multiple use cases which is great!

I'm not entirely sure what the end point to query this should be. in #116 we discussed adding a /history path or changing what is now /task (ListRunningTasks) to accept queries which might still be relevant even if that PR got stale.

**Comment by amanelis** [comment]
@stefanobaghino @keis I actually need the same thing. A way to specify a custom overwrite on the `task.Name` - If I follow correctly here, the piece that shows up in the Mesos UI under Eremetic as `Name` with the example of `Eremetic task 1gsVPCV6` would have the ability to now be overwritten.

@keis couldn't this just be opened up to the API? Similarly how @barryw and I did with adding the `privileged` flag  (cd1bd6b0697cd162827b4c09a9f710bdd2088208) to the request and then pushed onto the `task`?

**Comment by keis** [comment]
@amanelis Allowing the task name to be specified would be a first step towards solving the bigger issue here. Adding "name" as optional field to the request sounds fine.

PR's are most welcome!

**Comment by zmalik** [comment]
hey guys! can we divide this issue into two PRs? Following what @stefanobaghino said

- allowing non-unique `task.Name` to be provided by the user
- adding an HTTP endpoint to query status by `task.Name` 

It should be pretty easy the first PR. The second implies implementing in TaskDB interface a  `GetTasks(name string) ([]*Task, error)`

Anybody working on it? If not then I'm going to auto assign me that

**Comment by keis** [comment]
@zmalik yeah, splitting this up sounds good as allow the name makes sense as a feature on it's own.
